server.go

```diff
@@ -187,3 +224,196 @@ func copyWithHeader(r io.Reader, w io.Writer, header proto.Header) error {
+		// anyway since it will not be obvious they are in screen and doing
+		// things like switching windows makes mouse wheel scroll wonky due
+		// to the terminal doing the scrolling rather than screen itself
+		// (but again copy mode will work just fine).
+		"escape ^Ss",
+	}
+
+	dir := filepath.Join(os.TempDir(), "coder-screen")
+	config := filepath.Join(dir, "config")
+	screendir := filepath.Join(dir, "sockets")
+
+	err = os.MkdirAll(screendir, 0o700)
+	if err != nil {
+		return nil, xerrors.Errorf("unable to create %s for session %s: %w", screendir, id, err)
+	}
+
+	err = os.WriteFile(config, []byte(strings.Join(settings, "\n")), 0o644)
+	if err != nil {
+		return nil, xerrors.Errorf("unable to create %s for session %s: %w", config, id, err)
+	}
+
+	var sess *session
+	if rawSession, ok := sessions.Load(id); ok {
+		if sess, ok = rawSession.(*session); !ok {
+			return nil, xerrors.Errorf("found invalid type in session map for ID %s", id)
+		}
+	} else {
+		ctx, cancel := context.WithCancel(context.Background())
+		sess = &session{
+			ctx:   ctx,
+			ready: make(chan error, 1),
+		}
+		sessions.Store(id, sess)
+
+		// Starting screen with -Dm causes it to launch a server tied to this
+		// process, letting us attach to and kill it with the PID.
+		process, err := execer.Start(ctx, Command{
+			Command:    "screen",
+			Args:       append([]string{"-S", id, "-Dmc", config, command.Command}, command.Args...),
+			UID:        command.UID,
+			GID:        command.GID,
+			Env:        append(command.Env, "SCREENDIR="+screendir),
+			WorkingDir: command.WorkingDir,
+		})
+		if err != nil {
+			cancel()
+			err = xerrors.Errorf("failed to create session %s: %w", id, err)
+			sess.ready <- err
+			close(sess.ready)
+			return nil, err
+		}
+
+		sess.screenID = fmt.Sprintf("%d.%s", process.Pid(), id)
+
+		// Timeouts created with AfterFunc can be reset.
+		sess.timeout = time.AfterFunc(options.SessionTimeout, func() {
+			// Delete immediately in case it takes a while to clean up otherwise new
+			// connections with this ID will try to connect to the closing screen.
+			sessions.Delete(id)
+			// Close() will send a SIGTERM allowing screen to clean up.  If we kill it
+			// abruptly the socket will be left behind which is probably harmless but
+			// it causes screen -list to show a bunch of dead sessions (of course the
+			// user would not see this since we have a custom socket directory) and it
+			// seems ideal to let it clean up anyway.
+			process.Close()
+			select {
+			case <-sess.ctx.Done():
+				return
+			case <-time.After(5 * time.Second):
+				cancel() // Force screen to exit.
+			}
+		})
+
+		// Remove the session if screen exits.
+		go func() {
+			err := process.Wait()
+			// Screen exits with a one when it gets a SIGTERM.
+			if exitErr, ok := err.(ExitError); ok && exitErr.ExitCode() != 1 {
+				flog.Error("session %s exited with error %w", id, err)
+			}
+			cancel()
+			sess.timeout.Stop()
+			sessions.Delete(id)
```

Comment:
It's hard to understand the lifecycle of the session because the logic is divided here between the timeout function and this function --- is there a way to unify such that there is a single goroutine that `selects` and takes lifecycle actions?